File: 240929/마을 구분하기/seperate-village.py
n = int(input())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))
visited = []
for i in range(n):
    visited.append([0] * n)

village = 0
people = []

def in_range(x, y):
    if 0 <= x < n and 0 <= y < n:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] == 1 and not visited[x][y]:
        return True
    return False

def dfs(x, y):
    global num
    dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]

        if can_go(nx, ny):
            visited[nx][ny] = 1
            num += 1
            dfs(nx, ny)

num = 0
for row in range(n):
    for col in range(n):
        if area[row][col] == 1 and not visited[row][col]:
            village += 1
            num = 1
            visited[row][col] = 1
            dfs(row, col)
            people.append(num)

print(village)
people.sort()
for item in people:
    print(item)

File: 250222/마을 구분하기/seperate-village.cpp
#include <iostream>
#include <algorithm>
using namespace std;

#define endl '\n';

int n;
int area[25][25];
int village = 0;
int people[25] = {};
bool visited[25][25];

int dx[] = { -1, 0, 1, 0 };
int dy[] = { 0, 1, 0, -1 };

bool InRange(int x, int y) {
    return (0 <= x && x < n && 0 <= y && y < n);
}

bool CanGo(int x, int y) {
    return (InRange(x, y) && area[x][y] == 1 && !visited[x][y]);
}

void DFS(int x, int y, int village) {
    for (int dr = 0; dr < 4; dr++) {
        int nx = x + dx[dr];
        int ny = y + dy[dr];

        if (CanGo(nx, ny)) {
            people[village]++;
            visited[nx][ny] = true;
            DFS(nx, ny, village);
        }
    }
}

int main() {
    cin >> n;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> area[i][j];

    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            if (area[i][j] && !visited[i][j]) {
                people[village]++;
                visited[i][j] = true;
                DFS(i, j, village);
                village++;
            }
        }

    sort(people, people+village);
    cout << village << endl;
    for (int i = 0; i < village; i++)
        cout << people[i] << endl;
    return 0;
}